tuned_lens/scripts/train_loop.py

```diff
@@ -16,246 +18,362 @@
+            assert self.resume.is_dir()
+
+            print(f"Loading checkpoint from {self.resume}")
+            ddp_lens.load_state_dict(th.load(self.resume))
+
+        tokens_per_sample = len(data[0]["input_ids"])
+
+        grad_acc_steps = self.calculate_gradient_accumulation_steps(tokens_per_sample)
+
+        if self.sharding.cpu_offload and grad_acc_steps > 1:
+            raise ValueError("CPU offloading cannot be used with gradient accumulation")
+
+        metrics = defaultdict(list)
+        total_batches = self.num_steps * grad_acc_steps
+
+        pbar = tqdm(islice(dl, total_batches), desc="Training", total=total_batches)
+        for batch_idx, batch in enumerate(pbar, start=1):
+            assert isinstance(batch, dict)
+            batch = send_to_device(batch, th.device(local_rank()))
+            with th.autocast("cuda"):
+                output = model(**batch, output_hidden_states=True)
+
+            final_logits = output.logits
+            stream = ResidualStream(
+                embeddings=output.hidden_states[0], layers=output.hidden_states[1:-1]
+            )
+
+            shift = self.token_shift
+            if self.loss == "ce":
+                labels = batch["input_ids"]
+
+                # Predict the *next* token by default w/ cross entropy
+                if shift is None:
+                    shift = 1
+            elif self.loss == "kl":
+                labels = final_logits.log_softmax(dim=-1)
+
+                # Match the *current* token distribution by default
+                if shift is None:
+                    shift = 0
+            else:
+                raise NotImplementedError(f"Unknown loss {self.loss}")
+
+            labels = shift_labels(labels, shift)
+
+            # We do this sequentially to save VRAM
+            for i, (name, h) in enumerate(stream.items()):
+                # bfloat16 has larger dynamic range than float16 and seems to be better
+                # for computing log softmax & KL loss
+                with th.autocast("cuda", dtype=th.bfloat16):
+                    logits = shift_preds(ddp_lens(h, idx=i), shift)
+
+                    if self.loss == "ce":
+                        loss = th.nn.functional.cross_entropy(
+                            logits.flatten(0, -2), labels.flatten()
+                        )
+                    elif self.loss == "kl":
+                        loss = th.sum(
+                            labels.exp() * (labels - logits.log_softmax(-1)), dim=-1
+                        ).mean()
+                    else:
+                        raise NotImplementedError
+
+                    # Log the loss *before* LASSO regularization
+                    logging_loss = loss.detach()
+                    maybe_all_reduce(logging_loss)
+                    if local_rank() == 0:
+                        metrics[f"loss/{name}"].append(logging_loss)
+
+                    # Add sparsity regularizer
+                    if self.lasso:
```

Comment:
I have literally never used this and only included it because Igor did an experiment on it once so it can probably be removed